sql/server.go

```diff
@@ -106,13 +106,27 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Pick up current session state.
+	// The request user is validated in ServeHTTP. Even in insecure mode,
+	// it is guaranteed not to be empty.
+	planMaker := planner{user: args.GetUser()}
+	// TODO(tschottdorf) will have to validate the Session information (for
+	// instance, whether access to the stored database is permitted).
+	if err := gogoproto.Unmarshal(args.Session, &planMaker.session); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
 	// Send the Request for SQL execution and set the application-level error
-	// on the reply.
-	reply, err := s.exec(args)
-	if err != nil {
-		errProto := proto.Error{}
-		errProto.SetResponseGoError(err)
-		reply.Error = &errProto
+	// for each result in the reply.
+	reply := s.exec(args, &planMaker)
+
+	// Restore the session state even if there were application-level errors.
+	if bytes, err := gogoproto.Marshal(&planMaker.session); err != nil {
+		// Marshaling a `Session` never errors (known from reading the code).
+		panic(err)
```

Comment:
now that it's not so ugly to handle here, we can actually do proper error handling and write an http error. not required for this PR